File: python/Q424_longest_repeating_character_replacement.py
#
# @lc app=leetcode id=424 lang=python3
#
# [424] Longest Repeating Character Replacement
#


# @lc code=start
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        result: int = 0
        count: dict[str, int] = {}

        left: int = 0
        max_frequency: int = 0

        for right in range(len(s)):
            new_count: int = 1 + count.get(s[right], 0)

            count[s[right]] = new_count
            if new_count > max_frequency:
                max_frequency = new_count

            while (right - left + 1) - max_frequency > k:
                count[s[left]] -= 1
                left += 1

            if (current_length := right - left + 1) > result:
                result = current_length

        return result


# @lc code=end

File: rust/src/leetcode/Q424_longest_repeating_character_replacement.rs
/*
 * @lc app=leetcode id=424 lang=rust
 *
 * [424] Longest Repeating Character Replacement
 */

// @lc code=start
use std::collections::HashMap;

impl Solution {
    pub fn character_replacement(s: String, k: i32) -> i32 {
        let chars: Vec<char> = s.chars().collect();
        let mut current_map: HashMap<char, i32> = HashMap::new();
        let mut left: usize = 0;
        let mut max_frequency: i32 = 0;
        let mut result: usize = 0;

        for right in 0..chars.len() {
            *current_map.entry(chars[right]).or_insert(0) += 1;
            if current_map[&chars[right]] > max_frequency {
                max_frequency = current_map[&chars[right]];
            }

            while max_frequency + k < (right - left + 1) as i32 {
                *current_map.get_mut(&chars[left]).unwrap() -= 1;
                left += 1;
            }

            if right - left + 1 > result {
                result = right - left + 1;
            }
        }

        result as i32
    }
}
// @lc code=end
